gen/list_test.go

```diff
@@ -0,0 +1,220 @@
+//
+// FromWire
+//  Required list field
+//      wire.Value has an empty list field: decodes into empty.
+//      wire.Value is missing the list field: fails.
+//
+//  Optional list field
+//      wire.Value has an empty list field: decodes into empty.
+//      wire.Value is missing the list field: decodes to nil.
+//
+// We also test that full round trip works as expected with protocol en/decoding.
+
+// TestListRequired tests that empty slice and nil allocate an empty list after deserialization via
+// FromWire.
+func TestListRequiredToWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		p          *tc.ListOfRequiredPrimitives
+		offTheWire wire.Value
+		expect     []string
+	}{
+		{
+			"required nil list: wire.Value with empty list",
+			&tc.ListOfRequiredPrimitives{
+				ListOfStrings: nil,
+			},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{
+					ID:    1,
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			[]string{}, // we allocate a new slice if it doesn't exist.
+		},
+		{
+			"required empty list: wire.Value with empty list",
+			&tc.ListOfRequiredPrimitives{
+				ListOfStrings: []string{},
+			},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{
+					ID:    1,
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			[]string{},
+		},
+	}
+	for _, tt := range tests {
+		// required
+		w, err := tt.p.ToWire()
+		assert.NoError(t, err, "failed to serialize: %encodedValue", tt.p)
+		assert.True(t, wire.ValuesAreEqual(tt.offTheWire, w))
+
+		// Round trip them all.
+		freshestV, b := assertBinaryRoundTrip(t, w, tt.desc)
+		assert.True(t, b, "failed round trip")
+		assert.True(t, tt.p.Equals(tt.p))
+		// Allocate a new instance to serialize from Thrift representation.
+		x := new(tc.ListOfRequiredPrimitives)
+		if assert.NoError(t, x.FromWire(freshestV), tt.desc) {
+			assert.Equal(t, tt.expect, x.ListOfStrings)
+		}
+	}
+}
+
+func TestListRequiredFromWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		offTheWire wire.Value
+		expected   *tc.ListOfRequiredPrimitives
+	}{
+		{
+			"empty list field decodes into an empty slice",
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{
+					ID:    1,
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			&tc.ListOfRequiredPrimitives{
+				ListOfStrings: []string{},
+			},
+		},
+	}
+	for _, tt := range tests {
+		x := new(tc.ListOfRequiredPrimitives)
+		if assert.NoError(t, x.FromWire(tt.offTheWire), tt.desc) {
+			assert.Equal(t, tt.expected, x)
+			_, b := assertBinaryRoundTrip(t, tt.offTheWire, tt.desc)
+			assert.True(t, b, "failed round trip")
+		}
+	}
+}
+
+// Error if required list is missing in the wire representation.
+func TestListRequiredFromWireError(t *testing.T) {
```

Comment:
For this and TestListRequiredFromWire, since there's only one case each, we can do,

```
func TestListRequiredFromWire(t *testing.T) {
  t.Run("empty list decodes into an empty slice", func(t *testing.T) {
    give := wire.NewValueList(..)
    want := tc.ListOfPrimitives{...}
  
    var got tc.ListOfRequiredPrimitives
    require.NoError(t, got.FromWire(give), "failed to decode")
    assert.Equal(t, want, got, "decoded value did not match")
    
    _, ok := assertBinaryRoundTrip(..)
    assert.True(t, ok, "failed to round trip")
  })
  
  t.Run("absent list field results in an error", func(t *testing.T) {
    ...
  })
}
```